Update python binding for in-place foreach to return `List[Tensor]`

crcrpar opened the pull request:
fixes #104817 
taking over #118622

```c++
// _foreach_atan_
static PyObject * THPVariable__foreach_atan_(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_foreach_atan_(TensorList self)",
  }, /*traceable=*/false);

  ParsedArgs<1> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  // aten::_foreach_atan_(Tensor(a!)[] self) -> ()

  // auto dispatch__foreach_atan_ = [](at::TensorList self) -> at::TensorList {
  auto dispatch__foreach_atan_ = [](at::TensorList self) -> void {
    pybind11::gil_scoped_release no_gil;
    at::_foreach_atan_(self);
  };
  dispatch__foreach_atan_(_r.tensorlist(0));
  PyObject* self_tensorlist = _r.args[0];
  Py_INCREF(self_tensorlist);
  return self_tensorlist;
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
...
// _foreach_div_
static PyObject * THPVariable__foreach_div_(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_foreach_div_(TensorList self, ScalarList scalars)",
    "_foreach_div_(TensorList self, Tensor other)",
    "_foreach_div_(TensorList self, TensorList other)",
    "_foreach_div_(TensorList self, Scalar scalar)",
  }, /*traceable=*/false);

  ParsedArgs<2> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  switch (_r.idx) {
    case 0: {
      // aten::_foreach_div_.ScalarList(Tensor(a!)[] self, Scalar[] scalars) -> ()

      // auto dispatch__foreach_div_ = [](at::TensorList self, at::ArrayRef<at::Scalar> scalars) -> at::TensorList {
      auto dispatch__foreach_div_ = [](at::TensorList self, at::ArrayRef<at::Scalar> scalars) -> void {
        pybind11::gil_scoped_release no_gil;
        at::_foreach_div_(self, scalars);
      };
      dispatch__foreach_div_(_r.tensorlist(0), _r.scalarlist(1));
      PyObject* self_tensorlist = _r.args[0];
      Py_INCREF(self_tensorlist);
      return self_tensorlist;
    }
    case 1: {
      // aten::_foreach_div_.Tensor(Tensor(a!)[] self, Tensor other) -> ()

      // auto dispatch__foreach_div_ = [](at::TensorList self, const at::Tensor & other) -> at::TensorList {
      auto dispatch__foreach_div_ = [](at::TensorList self, const at::Tensor & other) -> void {
        pybind11::gil_scoped_release no_gil;
        at::_foreach_div_(self, other);
      };
      dispatch__foreach_div_(_r.tensorlist(0), _r.tensor(1));
      PyObject* self_tensorlist = _r.args[0];
      Py_INCREF(self_tensorlist);
      return self_tensorlist;
    }
    case 2: {
      // aten::_foreach_div_.List(Tensor(a!)[] self, Tensor[] other) -> ()

      // auto dispatch__foreach_div_ = [](at::TensorList self, at::TensorList other) -> at::TensorList {
      auto dispatch__foreach_div_ = [](at::TensorList self, at::TensorList other) -> void {
        pybind11::gil_scoped_release no_gil;
        at::_foreach_div_(self, other);
      };
      dispatch__foreach_div_(_r.tensorlist(0), _r.tensorlist(1));
      PyObject* self_tensorlist = _r.args[0];
      Py_INCREF(self_tensorlist);
      return self_tensorlist;
    }
    case 3: {
      // aten::_foreach_div_.Scalar(Tensor(a!)[] self, Scalar scalar) -> ()

      // auto dispatch__foreach_div_ = [](at::TensorList self, const at::Scalar & scalar) -> at::TensorList {
      auto dispatch__foreach_div_ = [](at::TensorList self, const at::Scalar & scalar) -> void {
        pybind11::gil_scoped_release no_gil;
        at::_foreach_div_(self, scalar);
      };
      dispatch__foreach_div_(_r.tensorlist(0), _r.scalar(1));
      PyObject* self_tensorlist = _r.args[0];
      Py_INCREF(self_tensorlist);
      return self_tensorlist;
    }
  }
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

crcrpar committed:
return self at python binding level

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
in-place return tensors now

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
assert self_arg

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
return TensorList

Without `wrap`, I see
```
error: no viable conversion from returned value of type 'at::TensorList' (aka 'ArrayRef<at::Tensor>') to function return type 'PyObject *' (aka '_object *')
      return dispatch__foreach_mul_(_r.tensorlist(0), _r.scalarlist(1));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make sure the ids are identical

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
avoid `wrap` and return `PyObject*`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1349,7 +1351,32 @@ def go(f: NativeFunction) -> str:
+  {dispatch_callee}({dispatch_args});
+}};
+dispatch_{name}({lambda_args}){set_requires_grad};
+PyObject* self_tensorlist = _r.args[0];
+Py_INCREF(self_tensorlist);
+return self_tensorlist;
```
As per https://github.com/pytorch/pytorch/pull/118622#discussion_r1509212846, I sort of optimistically return the first element of args.
@soulitzer would this follow your thought?

soulitzer replied:
That sounds good to me.

nit: btw I wonder if there's opportunity for some deduplication. The two paths seem to do the same things except for the return?

Skylion007 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1349,7 +1351,32 @@ def go(f: NativeFunction) -> str:
+                assert self_arg is not None and is_tensor_list_type(
+                    self_arg.argument.type
+                )
+                return f"""\
```
We not just use textwrap.dedent here instead to make this multiline fstring a little cleaner?

crcrpar committed:
deduplicate; def return stmt in if-else

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
revert unnecessary change: {lambda_return} -> void

it wouldn't matter a lot given this change is in `if lambda_return ==
"void"`.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer reviewed (approved):
LGTM

crcrpar commented:
@pytorchbot merge
